src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
+        Node<T> prev = next.prev;
+        Node<T> newNode = new Node<>(prev, element, next);
+        next.prev = newNode;
+        if (prev == null) {
+            head = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        size++;
+    }
+
+    private void setFirstValue(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        head = tail = newNode;
+        size++;
+    }
+
+    private void linkLast(T element) {
+        Node<T> newNode = new Node<>(tail, element, null);
+        tail.next = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private void unlink(Node<T> node) {
+        size--;
+        if (node == head && node == tail) {
+            head = null;
+            tail = null;
+            return;
+        } else if (node == head) {
+            head = node.next;
+            node.next.prev = null;
+            return;
+        } else if (node == tail) {
+            tail = node.prev;
+            node.prev.next = null;
+            return;
+        }
+        node.next.prev = node.prev;
+        node.prev.next = node.next;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            setFirstValue(value);
+            return;
+        }
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        addCheck(index);
+        if (isEmpty()) {
+            setFirstValue(value);
+            return;
+        }
+        if (index == size) {
+            linkLast(value);
+            return;
+        }
+        linkBefore(value, findNodeByIndex(index));
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new IndexOutOfBoundsException("List is null");
+        }
+        for (T item : list) {
+            this.add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> oldNode = findNodeByIndex(index);
+        T oldItem = oldNode.item;
+        oldNode.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = findNodeByIndex(index);
+        T value = node.item;
+        unlink(node);
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        try {
+            unlink(findNodeByValue(object));
+        } catch (RuntimeException e) {
+            return false;
+        }
+        return true;
```

Comment:
* first of all, should we catch runtime exceptions? It is a bad practice to do your logic with try\catch blocks instead of booleans and ifs, even though it is not the best either
* Second thing is that here having a separate method is causing more complexity, could have just get the logic from `findNodeByValue` and instead of returning the node, deleted it right away and returned true, and false otherwise, we can have it like this since we are not reusing it anywhere else at the moment